be/src/storage/tablet_updates.cpp

```diff
@@ -878,15 +878,35 @@ bool TabletUpdates::_check_status_msg(std::string_view msg) {
     return has_memory && (has_exceed_limit || has_alloc_failed);
 }
 
+bool TabletUpdates::_retry_times_limit() {
+    const size_t base_interval = config::retry_apply_interval_second;
+    const size_t max_interval = 600;
+    const size_t max_retries = max_interval / base_interval;
+    const size_t failed_retries = _apply_failed_time;
+
+    size_t total_duration = 0;
+
+    if (failed_retries <= max_retries) {
+        total_duration = failed_retries * (base_interval + base_interval * failed_retries) / 2;
+    } else {
+        const size_t sum_base = max_retries * (base_interval + base_interval * max_retries) / 2;
+        total_duration = sum_base + (failed_retries - max_retries) * max_interval;
+    }
+
+    return total_duration < config::retry_apply_timeout;
+}
+
 bool TabletUpdates::_is_retryable(Status& status) {
     switch (status.code()) {
     case TStatusCode::OK:
     case TStatusCode::MEM_LIMIT_EXCEEDED:
     case TStatusCode::MEM_ALLOC_FAILED:
     case TStatusCode::TIMEOUT:
         return true;
+    case TStatusCode::CORRUPTION:
+        return false;
     default:
-        return _check_status_msg(status.message());
+        return _check_status_msg(status.message()) || _retry_times_limit();
```

Comment:
does this mean the retry is always allowed if it is a MEM_LIMIT_EXCEEDED error no matter the total duration exceeds the configured time out limit or not.

Reply:
> does this mean the retry is always allowed if it is a MEM_LIMIT_EXCEEDED error no matter the total duration exceeds the configured time out limit or not.

Yes. If the error is due to insufficient memory, it will keep retrying. If it's due to corruption, it will not retry. For other errors with uncertain causes, it will attempt retries.

Reply:
so the timeout just takes effect for retry unknown/unhandled error. better add comments somewhere to explicit the use case.

Reply:
> so the timeout just takes effect for retry unknown/unhandled error. better add comments somewhere to explicit the use case.

done.